beep/structure.py

```diff
@@ -1235,39 +1235,38 @@ def determine_structuring_parameters(
         if run_parameter is not None:
             if {"capacity_nominal"}.issubset(run_parameter.columns.tolist()):
                 nominal_capacity = run_parameter["capacity_nominal"].iloc[0]
-            if {"discharge_cutoff_voltage", "charge_cutoff_voltage"}.issubset(
-                run_parameter.columns
-            ):
+            if {"discharge_cutoff_voltage", "charge_cutoff_voltage"}.issubset(run_parameter.columns):
                 v_range = [
                     all_parameters["discharge_cutoff_voltage"].min(),
                     all_parameters["charge_cutoff_voltage"].max(),
                 ]
-            if {
-                "diagnostic_type",
-                "diagnostic_start_cycle",
-                "diagnostic_interval",
-            }.issubset(run_parameter.columns):
+            if {"diagnostic_type", "diagnostic_start_cycle", "diagnostic_interval"}.issubset(run_parameter.columns):
                 if run_parameter["diagnostic_type"].iloc[0] == "HPPC+RPT":
                     hppc_rpt = ["reset", "hppc", "rpt_0.2C", "rpt_1C", "rpt_2C"]
                     hppc_rpt_len = 5
-                    diagnostic_starts_at = [
-                        1,
-                        1
-                        + run_parameter["diagnostic_start_cycle"].iloc[0]
-                        + 1 * hppc_rpt_len,
-                    ]
-                    for i in range(1, 100):
-                        diag_cycle_num = (
-                            i
-                            * (
-                                run_parameter["diagnostic_interval"].iloc[0]
-                                + hppc_rpt_len
-                            )
-                            + 1
-                            + run_parameter["diagnostic_start_cycle"].iloc[0]
-                            + 1 * hppc_rpt_len
-                        )
-                        diagnostic_starts_at.append(diag_cycle_num)
+                    initial_diagnostic_at = [1, 1 + run_parameter["diagnostic_start_cycle"].iloc[0] + 1 * hppc_rpt_len]
+                    # Calculate the number of steps present for each cycle in the diagnostic as the pattern for
+                    # the diagnostic. If this pattern of steps shows up at the end of the file, that indicates
+                    # the presence of a final diagnostic
+                    diag_0_pattern = [len(self.data[self.data.cycle_index == x].step_index.unique()) for x in
```

Comment:
@viveklam I think this will be able to find the final diagnostic in the file. It creates a list of the diagnostic cycles (reset, hppc, etc.) with the number of steps for each cycle. Then it looks at the last ~5 cycles in the file and if they have the same pattern of number of steps for those final cycles, it marks the those as belonging to a diagnostic.

Reply:
ie. diag_0_pattern = [2, 7, 2 ,2, 2]